src/riak_kv_vnode.erl

```
@@ -2367,12 +2380,31 @@ encode_and_put_no_sib_check(Obj, Mod, Bucket, Key, IndexSpecs, ModState,
                         false ->
                             ok
                     end,
-                    PutRet = Mod:put(Bucket, Key, IndexSpecs, EncodedVal,
-                                     ModState),
+                    case Sync of
+                        true ->
+                            PutRet = maybe_flush_put(Mod, Bucket, Key, IndexSpecs,
+                                                     EncodedVal, ModState);
+                        false ->
+                            PutRet = Mod:put(Bucket, Key, IndexSpecs, EncodedVal,
+                                             ModState)
+                    end,
                     {PutRet, EncodedVal}
             end
     end.
 
+-spec maybe_flush_put(Mod::term(), Bucket::riak_object:bucket(), Key::riak_object:key(),
+                      IndexSpecs::list(), EncodedVal::binary(), ModState::term()) ->
+           {{ok, UpdModState::term()}, EncodedObj::binary()} |
+           {{error, Reason::term(), UpdModState::term()}, EncodedObj::binary()}.
+maybe_flush_put(Mod, Bucket, Key, IndexSpecs, EncodedVal, ModState) ->
+    {ok, Capabilities} = Mod:capabilities(ModState),
+    case lists:member(flush_put, Capabilities) of
+        true ->
+            Mod:flush_put(Bucket, Key, IndexSpecs, EncodedVal, ModState);
+        false ->
```

I generally get the feeling that there is a lack of warnings. From an ops perspective, things seem to fail silently (e.g. pw=2 failing during replication to a sink site with two failed nodes). So I guess it would be good to flag this up. Any better ways to do it? I was worried about the spamminess.